### ISSUE TITLE
MemoryError for large search spaces in smb-optimizers

### ISSUE THREAD

**Issue by SimonBlanke** [SimonBlanke/Gradient-Free-Optimizers#3]
In SimonBlanke/Hyperactive#10 the user @gr8den reported a memory-error in the _DecisionTreeOptimizer_ for the following search space size (9857, 1620, 1620, 9857).

Error message:
`
MemoryError: Unable to allocate 1.81 PiB for an array with shape (9857, 1620, 1620, 9857) and data type int64
`

This error is expected for the current implementation of sequence-model-based-optimizers. However i will write a print into the code that warns the user if the search space exceeds a certain size (i will do tests to determine this size).

I would also like to explore how to reduce the memory- and computation-load of smb-optimizers in this issue.

**Comment by SimonBlanke** [comment]
This issue has been patched in multiple ways:
 - an added optimization parameter "warnings" for all smb-optimizers, which warns the user of the number of positions in the search space is too large: 6fd6414701fddb9e8bb7567deae99071521b7b86
 - an added optimization parameter "sampling" that reduces the number of positions (after they are generated) by random sampling: 807e2b830f278d09915b84ad3cc6a017f4657962
 - an added optimization algorithm "bayesian powell's method" that produces far less memory load: 167b211414d343cc344d4b1ac943c55f2c4eaffd

All of those added features will be in version 0.3 of Gradient-Free-Optimizers

**Comment by SimonBlanke** [comment]
I reopened this issue because of the sampling feature:
> an added optimization parameter "sampling" that reduces the number of positions (after they are generated) by random sampling: 807e2b8

This feature can help in the case when the search space is not so big that it overloads the memory, but large enough to help speed up the .predict-method of the surrogate model via sampling of the position list from the search space. 
I am not content with a solution for this special case. The random sampling should also protect against the memory overload. There are two different ways I could do that:
 - sample positions one by one from the search space and make sure no position repeats. (slow)
 - sample multiple positions at once with with possible repetitions. (fast)

In both cases we avoid the creation of the entire grid from the search space, which would prevent a memory overload. I will test both solutions.

**Comment by SimonBlanke** [comment]
I failed to recognize, that I already solved this issue some time ago.

I solved this problem with a combination of both solutions described before.

If the search-space is larger than 10000000 positions it gets reduced to this number by default. I selected this threshold, because an additional factor of 10 blows up the memory on my machine just by calculating all possible positions in the search space. So this seems to be a good upper limit. The search space is reduced here by just randomly sampling positions from each dimension separately. This first step is important, because it avoids problems if the search-space is extremely large.

Additionally the smbo will sample 1000000 (per default) randomly from those possible positions. This is done to avoid problems of the gaussian-process-regressor (or other surrogate-models) that would receive too many samples and blow up the memory.
